### PR TITLE
Document the sharing map, and move sharing map unit tests to catch

### PR DESCRIPTION
**Pull request by danpoe**

### CODE REVIEW

**Comment by tautschnig** [comment]
Build failures are genuine:
```
../src/util/sharing_map.h:486:15: error: unused variable 'c' [-Werror,-Wunused-variable]
```
It seems that necessitates some changes to `sharing_mapt`. While at that: would it be possible to please add some documentation that in particular covers 1) complexity of operations and 2) reference counting/garbage collection semantics?

**Comment by danpoe** [comment]
@tautschnig Ok, I'll add some documentation.

**Comment by hannes-steffenhagen-diffblue** [comment]
`unit/sharing_map.cpp` is currently excluded from the cmake build (`unit/CMakeLists.txt`, line 11). 

Also, the test for sharing_map and sharing_node should probably be under the unit/util directory.

File: unit/Makefile [hunk 1]
```diff
@@ -28,6 +28,7 @@ SRC += unit_tests.cpp \
        java_bytecode/inherited_static_fields/inherited_static_fields.cpp \
        pointer-analysis/custom_value_set_analysis.cpp \
        sharing_node.cpp \
+       sharing_map.cpp \
```
Comment (tautschnig) [inline unit/Makefile]: As @hannes-steffenhagen-diffblue said: these should likely go in the `util` folder.

File: src/util/sharing_map.h [hunk 2]
```diff
@@ -450,7 +450,7 @@ SHARING_MAPT2(, size_type)::erase(
     return 0;
 
   node_type *del=nullptr;
-  unsigned del_bit;
+  unsigned del_bit=0;
```
Comment (tautschnig) [inline src/util/sharing_map.h]: I'm pretty sure `unsigned` should not be used here at all and any such use should be a `std::size_t` instead.

File: src/util/sharing_map.h [hunk 3]
```diff
@@ -39,6 +39,66 @@ Author: Daniel Poetzl
...
+/// maps.
+///
+/// The map is implemented as a fixed-height n-ary trie. The height H and the
+/// maximum number of children per inner node S are determined by the two
+/// configuration parameters `bits` and `chunks` in sharing_map.h. It holds
+/// that H = `bits` / `chunks` and S = 2 ^ `chunks`.
+///
+/// When inserting a key-value pair into the map, first the hash of its key is
+/// computed. The `bits` number of lower order bits of the hash are deemed
+/// significant, and are grouped into `bits` / `chunk` chunks). The hash is then
```
Comment (tautschnig) [inline src/util/sharing_map.h]: The `)` does not have a matching opening parenthesis.

File: src/util/sharing_map.h [hunk 4]
```diff
@@ -39,6 +39,66 @@ Author: Daniel Poetzl
...
+/// treated as a string (with each chunk representing a character) for the
+/// purposes of determining the position of the key-value pair in the trie. The
+/// actual key-value pairs are stored in the leaf nodes. Collisions (i.e., two
+/// different keys yield the same "string"), are handled by chaining the
+/// corresponding key-value pairs in a `std::list`.
```
Comment (tautschnig) [inline src/util/sharing_map.h:62]: Would a `std::forward_list` be sufficient?

File: src/util/sharing_map.h [hunk 5]
```diff
@@ -39,6 +39,66 @@ Author: Daniel Poetzl
   CV typename sharing_mapt<keyT, valueT, hashT, predT>::ST \
   sharing_mapt<keyT, valueT, hashT, predT>
 
+/// A map implemented as a tree where subtrees can be shared between different
+/// maps.
```
Comment (tautschnig) [inline src/util/sharing_map.h]: Can you somewhere explain how to generate sharing (likely via copy and/or assignment operations)?

**Comment by tautschnig** [comment]
Is is possible to add `iterator` and `const_iterator` members, or actually provide the same interface that a `std::map` provides?

**Comment by danpoe** [comment]
A few differences between the interfaces of `sharing_mapt` and `std::map` are necessary I think. I'd like to avoid having methods return/take iterators for the reasons below.

`std::map` is implemented as an RB tree, consisting of nodes that also have pointers to their parents. An iterator thus just needs to keep a pointer to a node. In our tree the nodes do not have parent pointers. An iterator would thus need to keep a stack of parent pointers. Moreover, dereferencing a `std::map` iterator yields a `std::pair`, which matches the implementation of `std::map` which stores key-value pairs as `std::pair`'s internally, whereas `sharing_mapt` does not store them as `std::pair`'s.

Since an iterator would need to contain a stack, it would cause unnecessary overhead to return an iterator on e.g. a call to `find()`. The iterator would also be less useful than a `std::map` iterator:

- The elements are not stored in any particular order in `sharing_mapt`, thus idioms like first retrieving an iterator with `find()` and then starting to iterate from that position are not meaningful.
- Passing an iterator to e.g. `insert()` in addition to a key as a hint would not be more efficient than just using the key, as when the map is modified all nodes on the path from the root of the tree to the newly inserted element need to be copied if they have `use_count > 1` (thus a traversal from the root to the insert position needs to be done in any case).
- Using an `iterator` (i.e. not a `const_iterator`) to iterate over `sharing_mapt` would break all the sharing as it would copy all the shared nodes on the assumption that they may be modified which is likely undesirable.

File: src/util/sharing_map.h (hunk 4 above)
Comment (danpoe) [reply src/util/sharing_map.h:62]: Yes, this is fixed in an upcoming PR.

File: src/util/sharing_map.h (hunk 5 above)
Comment (danpoe) [reply src/util/sharing_map.h]: Yes, I'll add an explanation for that.

**Review by hannes-steffenhagen-diffblue — APPROVED** [review]
Looks good to me.

**Comment by danpoe** [comment]
@tautschnig Would you mind if we merge this PR? (it's documentation, plus unit test refactoring, and some smaller fixes). We can move potential further discussions to #2126 or another upcoming PR.

**Comment by tautschnig** [comment]
Looking.

**Review by tautschnig — APPROVED** [review]

**Pull request merged**